hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DatanodeAdminMonitorImpl.java

```diff
@@ -168,6 +232,55 @@ public int getTrackedNodeCount() {
     return trackedNodes.size();
   }
 
+  synchronized void setMetricsToGauge() {
+    metrics.setTrackedContainersUnhealthyTotal(unhealthyContainers);
+    metrics.setTrackedRecommissionNodesTotal(trackedRecommission);
+    metrics.setTrackedDecommissioningMaintenanceNodesTotal(
+            trackedDecomMaintenance);
+    metrics.setTrackedContainersUnderReplicatedTotal(
+            underReplicatedContainers);
+    metrics.setTrackedContainersSufficientlyReplicatedTotal(
+            sufficientlyReplicatedContainers);
+    metrics.setTrackedPipelinesWaitingToCloseTotal(pipelinesWaitingToClose);
+    for (Map.Entry<String, ContainerStateInWorkflow> e :
+            containerStateByHost.entrySet()) {
+      metrics.metricRecordOfContainerStateByHost(e.getKey(),
+          e.getValue().sufficientlyReplicated,
+          e.getValue().underReplicatedContainers,
+          e.getValue().unhealthyContainers,
+          e.getValue().pipelinesWaitingToClose);
+    }
+  }
+
+  void resetContainerMetrics() {
+    pipelinesWaitingToClose = 0;
+    sufficientlyReplicatedContainers = 0;
+    unhealthyContainers = 0;
+    underReplicatedContainers = 0;
+
+    for (Map.Entry<String, ContainerStateInWorkflow> e :
```

Comment:
At the moment, to reset things, you need to iterate this map and clear all values. Then iterate it again and remove any entries that are no longer tracked.

What if, you just clear the map, which is a one liner.

Then make  ContainerStateInWorkflow a public inner class, and pass the `Map<String, ContainerStateInWorkflow>` directly to the metrics class and have it replace all its metrics internally with the new map.

That way reset becomes a lot easier too.

Also, if we add another metric for a host, we don't need to add another parameter to the `metricRecordOfContainerStateByHost` method, as it just receives the wrapper object anyway.

Reply:
Does the implementation as it stands now, protect against incomplete intermediate metrics? The metrics are snapshot via the call to getMetrics, but the metrics are set over several calls from the Decommission monitor to the metrics class and there is no synchronisation. Could we not set `trackedPipelinesWaitingToCloseTotal` and then before `trackedContainersUnderReplicatedTotal` is set, `getMetrics` is called giving an inconsistent result?

Probably, getMetrics() and any setters need synchronized, and even then you need to set everything in a single synchronized call.

We can build up a `Map<String, ContainerStateInWorkflow>` for each iteration and at the end of the iteration pass it to the metrics object like new `Map<>(mapJustBuiltUp)` - then you can clear the original and the Map and ContainerStateInWorkflow objects are referenced only in the new metrics class and will not be changed again.

Replace them all on the next call and then we don't need to worry about expiring individual nodes.

Reply:
On each scheduled run of the monitor, the implementation captures the current workflow state completely prior to flushing the metric update to the `NodeDecommissionMetric` object (`DatanodeAdminMonitorImpl.setMetricsToGauge`).  In doing so, it tries to keep each metric refresh pull from `getMetrics `display a full snapshot of the last captured run of the monitor.

> We can build up a Map<String, ContainerStateInWorkflow> for each iteration and at the end of the iteration pass it to the metrics object like new Map<>(mapJustBuiltUp)

Yes, I currently have been coding something just like that based on your earlier comment.  With this, the `Map<String, ContainerStateInWorkflow>` is passed to the `NodeDecommisionMetric.metricRecordOfContainerStateByHost`.  Within it, it sets the internal metrics by host and _also_ removes stale nodes that exited the workflow since the last run of the monitor snapshot captured.  It looks like what we are discussing.

Reply:
> On each scheduled run of the monitor, the implementation captures the current workflow state completely prior to flushing the metric update to the NodeDecommissionMetric object (DatanodeAdminMonitorImpl.setMetricsToGauge). In doing so, it tries to keep each metric refresh pull from getMetrics display a full snapshot of the last captured run of the monitor.

The code in `DatanodeAdminMonitorImpl.setMetricsToGauge` is:

```
synchronized void setMetricsToGauge() {
    metrics.setTrackedContainersUnhealthyTotal(unhealthyContainers);
    metrics.setTrackedRecommissionNodesTotal(trackedRecommission);
    metrics.setTrackedDecommissioningMaintenanceNodesTotal(
            trackedDecomMaintenance);
    metrics.setTrackedContainersUnderReplicatedTotal(
            underReplicatedContainers);
    metrics.setTrackedContainersSufficientlyReplicatedTotal(
            sufficientlyReplicatedContainers);
    metrics.setTrackedPipelinesWaitingToCloseTotal(pipelinesWaitingToClose);
    for (Map.Entry<String, ContainerStateInWorkflow> e :
...
```

It makes multiple calls to `metrics` and there is nothing stopping getMetrics() being called on the metrics object by another thread half way through the execution of `setMetricsToGauge`. This means the metrics can still be snapshot inconsistently.

To make it consistent, you need to synchronize in the metrics object and set ALL the metrics in a single synchronized call.

Reply:
> To make it consistent, you need to synchronize in the metrics object and set ALL the metrics in a single synchronized call.

Will do.  At least try to keep possible inconsistency to a minimum.  We won't report metrics that show one value in one sample, that clear, then go back to value due to our sampling from the monitor.

Reply:
Is this something we should look to add to the code, adding in a single call to the metrics object for all collected metrics to update?  Modify `DatanodeAdminMonitorImpl.setMetricsToGauge`?

Reply:
Latest commit contains the modifications we are discussing - 

> We can build up a Map<String, ContainerStateInWorkflow> for each iteration and at the end of the iteration pass it to the metrics object like new Map<>(mapJustBuiltUp)

Each iteration in the monitor collects snapshot of node metrics in workflow within threadLocal variables.

> Then make ContainerStateInWorkflow a public inner class, and pass the Map<String, ContainerStateInWorkflow> directly to the metrics class and have it replace all its metrics internally with the new map.

The `NodeDecommissionMetrics` uses the` ContainerStateInWorkflow` for each node to set the metric gauges pulled by `getMetrics`.  Each iteration clears the internal maps.